File: SampleProject/Assets/Scripts/Grid/C++/Grid.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;
using Random = System.Random;

namespace Grid
{
    /// <summary>
    /// TODO: Move to C++
    /// </summary>
    public struct AStartTypeInfo
    {
        public Grid typeGrid;
        public int[] useableTypes;
    }
    
    public class Grid
    {
        public int Width { get; }
        public int Height { get; }
        public int DefaultValue { get; }
        
        public int OutOfBoundsValue { get; }
        
        private readonly int[,] _grid;

        public int this[int x, int y]
        {
            get => _grid[x, y];
            set => _grid[x, y] = value;
        }
        
        public Grid(int width, int height, int defaultValue = -1, int outOfBoundsValue = Int32.MinValue)
        {
            Width = width;
            Height = height;
            OutOfBoundsValue = outOfBoundsValue;
            DefaultValue = defaultValue;

            _grid = new int[width,height];
            for (var i = 0; i < width; i++)
            {
                for (var j = 0; j < height; j++)
                {
                    _grid[i, j] = defaultValue;
                }
            }
        }

        public bool IsPointSet(Point coord) => _grid[coord.X, coord.Y] != DefaultValue;

        public Point GetRandomPointOfType(int type)
        {
            var pointsOfType = new List<Point>();
            
            for (var i = 0; i < Width; i++)
            {
                for (var j = 0; j < Height; j++)
                {
                    if (_grid[i,j] != type) continue;
                    pointsOfType.Add(new Point(i,j));
                }
            }
            var rand = new Random();
            return pointsOfType[rand.Next(0, pointsOfType.Count - 1)];
        }

        public List<Point> GetAdjacentPoints(Point point)
        {
            var adjacentCells = new List<Point>();
            if (point.X > 0)
            {
                adjacentCells.Add(new Point(point.X - 1,point.Y));
            }
            if (point.X < Width -1)
            {
                adjacentCells.Add(new Point(point.X + 1, point.Y));
            }
            if (point.Y > 0)
            {
             adjacentCells.Add(new Point(point.X,point.Y - 1));   
            }
            if (point.Y < Height -1)
            {
                adjacentCells.Add(new Point(point.X,point.Y + 1));   
            }
            return adjacentCells;
        }

        public List<Point> GetAdjacentPointsOfType(Point point, int type)
        {
            var adjacentPoints = GetAdjacentPoints(point);
            for (var i = adjacentPoints.Count - 1; i >= 0; i--)
            {
                    if (PointIsOfType(point, type)) continue;
                    adjacentPoints.RemoveAt(i);
            }
            return adjacentPoints;
        }
        
        private List<Point> GetAdjacentPointsOfTypes(Point point, int[] acceptedTypes)
        {
            var adjacentPoints = GetAdjacentPoints(point);
            for (var i = adjacentPoints.Count - 1; i >= 0; i--)
            {
                var isOfType = acceptedTypes.Any(type => PointIsOfType(adjacentPoints[i], type));
                if (isOfType) continue;
                adjacentPoints.RemoveAt(i);
            }
            return adjacentPoints;
        }

        private bool PointIsOfType(Point point, int type) => _grid[point.X, point.Y] == type;
        
        /// <summary>
        ///  
        /// </summary>
        /// <param name="point"> Point to get neighbors of</param>
        /// <returns>Array [Left neighbor, Top neighbor, Right neighbor, Down neighbor]</returns>
        public int[] GetAdjacentPointTypes(Point point)
        {
            var adjacentTypes = new [] {OutOfBoundsValue, OutOfBoundsValue, OutOfBoundsValue, OutOfBoundsValue};
            if (point.X > 0)
            {
                adjacentTypes[0] = _grid[point.X - 1, point.Y];
            }
            if (point.Y < Height - 1)
            {
                adjacentTypes[1] = _grid[point.X, point.Y + 1];
            }
            if (point.X < Width - 1)
            {
                adjacentTypes[2] = _grid[point.X + 1, point.Y];
            }
            if (point.Y > 0)
            {
                adjacentTypes[3] = _grid[point.X, point.Y - 1];
            }
            

            return adjacentTypes;
        }

        public List<Point> AStarSearchWithoutCost(Point startPoint, Point endPoint, AStartTypeInfo? typeInfo = null) =>
            AStarSearch(startPoint, endPoint, false, typeInfo);

        public List<Point> AStarSearchWithCost(Point startPoint, Point endPoint, AStartTypeInfo? typeInfo = null)
            => AStarSearch(startPoint, endPoint, true, typeInfo);

        private List<Point> AStarSearch(Point startPoint, Point endPoint, bool useCost, AStartTypeInfo? typeInfo = null)
        {
            var path = new List<Point>();

            var pointsToCheck = new List<Point>();
            var costDictionary = new Dictionary<Point,int>();
            var priorityDictionary = new Dictionary<Point, float>();
            var parentsDictionary = new Dictionary<Point,Point>();
            
            pointsToCheck.Add(startPoint);
            costDictionary.Add(startPoint, 0);
            priorityDictionary.Add(startPoint, 0);
            parentsDictionary.Add(startPoint, null);

            while (pointsToCheck.Count > 0)
            {
                var currentPoint = GetClosestPoint(pointsToCheck, priorityDictionary);
                pointsToCheck.Remove(currentPoint);
                Debug.Log(currentPoint.ToString() + endPoint);
                if (currentPoint.Equals(endPoint))
                {
                    path = GeneratePath(parentsDictionary, currentPoint);
                    return path;
                }

                var neighbors = typeInfo == null 
                    ? GetAdjacentPoints(currentPoint) 
                    : typeInfo.Value.typeGrid.GetAdjacentPointsOfTypes(currentPoint, typeInfo.Value.useableTypes);
                foreach (var neighbor in neighbors)
                {
                    var newCost = useCost 
                        ? costDictionary[currentPoint] + _grid[neighbor.X, neighbor.Y]
                        : 0;
                    if (costDictionary.ContainsKey(neighbor) && !(newCost < costDictionary[neighbor])) continue;
                    costDictionary[neighbor] = newCost;

                    var priority = newCost + ManhattanDistance(neighbor, endPoint);
                    pointsToCheck.Add(neighbor);
                    priorityDictionary[neighbor] = priority;

                    parentsDictionary[neighbor] = currentPoint;
                }
            }
            return path;
        }

        private float ManhattanDistance(Point point, Point endPoint) =>
            Math.Abs(endPoint.X - point.X) + Math.Abs(endPoint.Y - point.Y);

        private Point GetClosestPoint(List<Point> pointsToCheck, Dictionary<Point, float> distanceMap)
        {
            var candidate = pointsToCheck[0];
            foreach (var point in pointsToCheck.Where(point => distanceMap[point] < distanceMap[candidate]))
            {
                candidate = point;
            }

            return candidate;
        }
        
        private List<Point> GeneratePath(Dictionary<Point,Point> parentsDictionary, Point endPoint)
        {
            var path = new List<Point>();
            var parent = endPoint;
            while (parent != null && parentsDictionary.ContainsKey(parent))
            {
                path.Add(parent);
                parent = parentsDictionary[parent];
            }

            return path;
        }
    }
}

File: Grid/Grid/Grid.cpp
#include "pch.h"
#include "Grid.h"


Grid::Grid(int width, int height, int defaultValue, int outOfBoundsValue)
	:Width(width)
	,Height(height)
	,DefaultValue(defaultValue)
	,OutOfBoundsValue(outOfBoundsValue)
{
	for (int i = 0; i < width*height; i++)
	{
		m_Grid[i] = defaultValue;
	}
}

Grid::Grid(const Grid &g)
	:Width(g.Width)
	,Height(g.Height)
	,DefaultValue(g.DefaultValue)
	,OutOfBoundsValue(g.OutOfBoundsValue)
{
	for (int i = 0; i < g.Width* g.Height; i++)
	{
		m_Grid[i] = g.DefaultValue;
	}
}

int Grid::GetGridContent(Coordinate cooridnate)
{
	int pos = CoordinateToGridIdx(cooridnate);
	return m_Grid[pos];
}

void Grid::SetGridContent(Coordinate cooridnate, int value)
{
	int pos = CoordinateToGridIdx(cooridnate);
	m_Grid[pos] = value;
}

bool Grid::IsPositionSet(Coordinate cooridnate)
{
	int pos = CoordinateToGridIdx(cooridnate);
	return m_Grid[pos] != DefaultValue;
}

Coordinate Grid::GetRandomCooridanteOfValue(int value)
{
	std::vector<Coordinate> coordinatesOfType;
	for (int i = 0; i < Width * Height; i++)
	{
		if (m_Grid[i] == value)
		{
			Coordinate coordinate = GridIdxToCoordinate(i);
			coordinatesOfType.push_back(coordinate);
		}
	}

	int rand = (std::rand() % (coordinatesOfType.size() + 1));
	return coordinatesOfType[rand];
}

int Grid::GetAdjacentValidCoordinatesCount(Coordinate coordinate)
{
	int retval = 0;
	if (coordinate.X > 0)
	{
		retval++;
	}
	if (coordinate.X < Width - 1)
	{
		retval++;
	}
	if (coordinate.Y > 0)
	{
		retval++;
	}
	if (coordinate.Y < Height - 1)
	{
		retval++;
	}
	return retval;
}

std::vector<Coordinate> Grid::GetAdjacentVaildCoordinates(Coordinate coordinate)
{
	std::vector<Coordinate> retVal;
	if (coordinate.X > 0)
	{
		retVal.push_back({ coordinate.X - 1, coordinate.Y });
	}
	if (coordinate.X < Width - 1)
	{
		retVal.push_back({ coordinate.X + 1, coordinate.Y });
	}
	if (coordinate.Y > 0)
	{
		retVal.push_back({ coordinate.X, coordinate.Y - 1 });
	}
	if (coordinate.Y < Height - 1)
	{
		retVal.push_back({ coordinate.X, coordinate.Y + 1 });
	}
	return retVal;
}

std::vector<Coordinate> Grid::GetAdjacentValidCoordinatesWithValue(Coordinate coordinate, std::vector<int> value)
{
	std::vector<Coordinate> retVal;
	if (coordinate.X > 0)
	{
		Coordinate neighbor = { coordinate.X - 1, coordinate.Y };
		int gridIdx = CoordinateToGridIdx(neighbor);
		if (std::find(value.begin(), value.end(), m_Grid[gridIdx]) != value.end())
		{
			retVal.push_back(neighbor);
		}
	}
	if (coordinate.X < Width - 1)
	{
		Coordinate neighbor = { coordinate.X + 1, coordinate.Y };
		int gridIdx = CoordinateToGridIdx(neighbor);
		if (std::find(value.begin(), value.end(), m_Grid[gridIdx]) != value.end())
		{
			retVal.push_back(neighbor);
		}
	}
	if (coordinate.Y > 0)
	{
		Coordinate neighbor = { coordinate.X, coordinate.Y - 1};
		int gridIdx = CoordinateToGridIdx(neighbor);
		if (std::find(value.begin(), value.end(), m_Grid[gridIdx]) != value.end())
		{
			retVal.push_back(neighbor);
		}
	}
	if (coordinate.Y < Height - 1)
	{
		Coordinate neighbor = { coordinate.X, coordinate.Y + 1};
		int gridIdx = CoordinateToGridIdx(neighbor);
		if (std::find(value.begin(), value.end(), m_Grid[gridIdx]) != value.end())
		{
			retVal.push_back(neighbor);
		}
	}
	return retVal;
}

std::vector<int> Grid::GetAdjacentValidValues(Coordinate coordinate)
{
	std::vector<int> retVal;
	if (coordinate.X > 0)
	{
		int gridIdx = CoordinateToGridIdx({ coordinate.X + 1, coordinate.Y });
		retVal.push_back(m_Grid[gridIdx]);
	}
	if (coordinate.X < Width - 1)
	{
		int gridIdx = CoordinateToGridIdx({ coordinate.X - 1, coordinate.Y });
		retVal.push_back(m_Grid[gridIdx]);
	}
	if (coordinate.Y > 0)
	{
		int gridIdx = CoordinateToGridIdx({ coordinate.X, coordinate.Y - 1});
		retVal.push_back(m_Grid[gridIdx]);
	}
	if (coordinate.Y < Height - 1)
	{
		int gridIdx = CoordinateToGridIdx({ coordinate.X, coordinate.Y + 1});
		retVal.push_back(m_Grid[gridIdx]);
	}
	return retVal;
}

std::vector<Coordinate> Grid::AStarSearch(Coordinate start, Coordinate end, bool useCost)
{
	std::vector<Coordinate> path;

	std::vector<Coordinate> coordsToCheck;
	std::map<Coordinate, int> costMap;
	std::map<Coordinate, float> priorityMap;
	std::map<Coordinate, Coordinate> parentsMap;

	coordsToCheck.push_back(start);
	costMap[start] = 0;
	priorityMap[start] = 0;
	parentsMap[start] = { OutOfBoundsValue,OutOfBoundsValue };

	while (coordsToCheck.size() > 0)
	{
		auto current = GetClosestCoordinate(coordsToCheck, priorityMap);
		auto it = GetCurrentIterator(coordsToCheck, current);
		coordsToCheck.erase(it);
		
		if (current == end)
		{
			path = GeneratePath(parentsMap, current);
			return path;
		}

		auto neighbors = GetAdjacentVaildCoordinates(current);
		for (auto neighbor : neighbors)
		{
			auto newCost = useCost
				? costMap[current] + m_Grid[CoordinateToGridIdx(neighbor)]
				: 0;

			if (costMap.count(neighbor) && !(newCost < costMap[neighbor])) continue;
			costMap[neighbor] = newCost;

			auto priority = newCost + ManhattanDistance(neighbor, end);
			coordsToCheck.push_back(neighbor);
			priorityMap[neighbor] = priority;

			parentsMap[neighbor] = current;
		}
	}
	return path;
}

std::vector<Coordinate> Grid::AStarSearch(Coordinate start, Coordinate end, bool useCost, AStarTypeInfo typeInfo)
{
	std::vector<Coordinate> path;

	std::vector<Coordinate> coordsToCheck;
	std::map<Coordinate, int> costMap;
	std::map<Coordinate, float> priorityMap;
	std::map<Coordinate, Coordinate> parentsMap;

	coordsToCheck.push_back(start);
	costMap[start] = 0;
	priorityMap[start] = 0;
	parentsMap[start] = { OutOfBoundsValue,OutOfBoundsValue };

	if (std::find(typeInfo.UseableTypes.begin(), typeInfo.UseableTypes.end(), typeInfo.TypeGrid[CoordinateToGridIdx(start)]) != typeInfo.UseableTypes.end()) return path;

	while (coordsToCheck.size() > 0)
	{
		auto current = GetClosestCoordinate(coordsToCheck, priorityMap);
		auto it = GetCurrentIterator(coordsToCheck, current);
		coordsToCheck.erase(it);

		if (current == end)
		{
			path = GeneratePath(parentsMap, current);
			return path;
		}

		auto neighbors = typeInfo.TypeGrid->GetAdjacentValidCoordinatesWithValue(current, typeInfo.UseableTypes);
		for (auto neighbor : neighbors)
		{
			auto newCost = useCost
				? costMap[current] + m_Grid[CoordinateToGridIdx(neighbor)]
				: 0;

			if (costMap.count(neighbor) && !(newCost < costMap[neighbor])) continue;
			costMap[neighbor] = newCost;

			auto priority = newCost + ManhattanDistance(neighbor, end);
			coordsToCheck.push_back(neighbor);
			priorityMap[neighbor] = priority;

			parentsMap[neighbor] = current;
		}
	}
	return path;
}

int Grid::CoordinateToGridIdx(Coordinate cell)
{
	return cell.Y * Width + cell.X;
}

Coordinate Grid::GridIdxToCoordinate(int pos)
{
	return { pos / Width,pos % Width };
}

Coordinate Grid::GetClosestCoordinate(std::vector<Coordinate> coordsToChek, std::map<Coordinate, float> distanceMap)
{
	Coordinate candidate = coordsToChek[0];

	for (auto &coord : coordsToChek)
	{
		if (distanceMap[coord] < distanceMap[candidate])
		{
			candidate = coord;
		}
	}
	return candidate;
}

std::vector<Coordinate>::iterator Grid::GetCurrentIterator(std::vector<Coordinate> vector, Coordinate current)
{
	auto retVal = vector.begin();

	for (auto& coord : vector)
	{
		if (coord == current)
		{
			return retVal;
		}
		retVal++;
	}
	return vector.begin();
}

std::vector<Coordinate> Grid::GeneratePath(std::map<Coordinate, Coordinate> parentsMap, Coordinate end)
{
	std::vector<Coordinate> path;
	Coordinate parent = end;
	while (parent != Coordinate{OutOfBoundsValue, OutOfBoundsValue} && parentsMap.count(parent))
	{
		path.push_back(parent);
		parent = parentsMap[parent];
	}
	return path;
}

float Grid::ManhattanDistance(Coordinate current, Coordinate end)
{
	return fabs(end.X - current.X) + fabs(end.Y - current.Y);
}
